src/vsr/replica.zig

```diff
@@ -1550,6 +1566,17 @@ pub fn ReplicaType(
                 .realtime = realtime,
             };
 
+            // self.op includes requests in the pipeline_queue, we must explicitly account for
+            // requests in the request_queue to decide whether to accept this request.
+            if (self.op + self.pipeline.queue.request_queue.count + 1 > self.op_prepare_max()) {
+                log.debug("{}: on_request: ignoring op={} (too far ahead, prepare_max={}).", .{
+                    self.replica,
+                    self.op + self.pipeline.queue.request_queue.count + 1,
+                    self.op_prepare_max(),
+                });
+                defer self.message_bus.unref(request.message);
+                return;
+            }
```

Comment:
Hm, this looks a bit strange: my mental model is that request queue is a just a buffer where  we can/should add a new request if we can't otherwise execute it. 

That is, the case where we get a request, but we can't immediately convert it to prepare, is I think _exactly_ the case where we let the request sit in the request queue.

Reply:
My understanding about the request queue is similar, I understood at it as an overflow for the prepare queue! However, the reason why we can't admit requests into the prepare/queue freely in _on_request_ now is:
 
By virtue of this change, replicas that are past `vsr.Checkpoint.trigger_for_checkpoint(C)` (and have set `self.op_checkpoint() = C`) could still have their `self.op_prepare_max` at `vsr.Checkpoint.prepare_max_for_checkpoint(C-1)`, which is merely a pipeline_prepare_queue_max away. Freely admitting requests into the prepare/request queue could get to a primary to a point where it hits [this assert](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L1645-L1655) in _on_prepare_, which we're _aren't_ supposed to hit as a primary.

Here's a trace from failed VOPR run (without the admission check in on_request) that shows the issue more clearly. I've logged the sizes of the prepare & request queues, self.op, and op_prepare_max:
```
[debug] (replica): 1:: commit_op: popped prepare with type get_account_transfers op 26:: prepare queue 3/4, request queue 2/3, self.op 29, op_prepare_max 31
[debug] (replica): 1:: on_request:: popped request query_accounts: prepare queue 3/4, request queue 1/3, self.op 29, op_prepare_max 31
[debug] (replica): 1:: primary_repair_prepare:: pushed prepare query_accounts: prepare queue 4/4, request queue 1/3, self.op 29, op_prepare_max 31
[debug] (replica): 1:: commit_op: popped prepare with type create_transfers op 27:: prepare queue 3/4, request queue 1/3, self.op 30, op_prepare_max 31
[debug] (replica): 1:: on_request:: popped request lookup_transfers: prepare queue 3/4, request queue 0/3, self.op 30, op_prepare_max 31
[debug] (replica): 1:: primary_repair_prepare:: pushed prepare lookup_transfers: prepare queue 4/4, request queue 0/3, self.op 30, op_prepare_max 31
[debug] (replica): 1:: on_request:: pushed request create_transfers: prepare queue 4/4, request queue 1/3, self.op 31, op_prepare_max 31
[debug] (replica): 1:: on_request:: pushed request create_transfers: prepare queue 4/4, request queue 2/3, self.op 31, op_prepare_max 31
[debug] (replica): 1:: commit_op: popped prepare with type create_transfers op 28:: prepare queue 3/4, request queue 2/3, self.op 31, op_prepare_max 31
[debug] (replica): 1:: on_request:: popped request create_transfers: prepare queue 3/4, request queue 1/3, self.op 31, op_prepare_max 31
[debug] (replica): 1:: primary_repair_prepare:: pushed prepare create_transfers: prepare queue 4/4, request queue 1/3, self.op 31, op_prepare_max 31
```

We can see from the logs that even after we've committed op=27 (which is the checkpoint trigger) and op=28, our op_prepare_max is still at 31. When we reach self.op=31, we have capacity for two more requests in the request queue, which we would not be able to accommodate till we receive a replication quorum for the checkpoint and advance our op_prepare_max.

Reply:
>Freely admitting requests into the prepare/request queue could get to a primary to a point where it hits [this assert](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L1645-L1655) in on_prepare, which we're aren't supposed to hit as a primary.

I'd say that the problem here is not that we admit a request to request queu, but rather than we promote it to a prepare

Eg, here:

```
            if (self.status == .normal and self.primary()) { to 
                if (self.pipeline.queue.pop_request()) |request| {
                    // Start preparing the next request in the queue (if any).
                    self.primary_pipeline_prepare(request);
                }
```

I think we should add an if to check if we _can_ start preparing that op, and, correspondingly, add a call to `pop_request` to the place where we advance `op_prepare_max` on_ping

Reply:
Agreed, I had also explored the option of changing this in _commit_op_client_replies_ready_callback_, but realised that there are also certain requests that are implicitly sent ([send_pulse_to_self](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L9711), [send_request_upgrade_to_sef](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L9722)) which currently assume that there is always enough space in the prepare_queue (and consequently the WAL). This would spread out the code to check if there's enough space in the WAL over multiple functions, which is why I thought that containing it in _on_request_ be cleaner in comparison. Thoughts?

Reply:
>             if (self.status == .normal and self.primary()) { to 
>                if (self.pipeline.queue.pop_request()) |request| {
>                   // Start preparing the next request in the queue (if any).
>                   self.primary_pipeline_prepare(request);
>                }
>  I think we should add an if to check if we can start preparing that op, and, correspondingly, add a call to > pop_request to the place where we advance op_prepare_max on_ping

Also, I recall trying the approach of not popping the request from the request queue if we don't have enough space in the WAL (in [commit_op_client_replies_ready_callback](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L3870-L3873)). I believe we then end up violating [this](https://github.com/tigerbeetle/tigerbeetle/blob/main/src/vsr/replica.zig#L10319-L10321) invariant, since we pop a request from the prepare queue but we may not move one over from the request queue promptly.
```C
/// - If request_queue is not empty, then prepare_queue is full OR 1-less than full.
///   (The caller is responsible for maintaining this invariant. If the caller removes an entry
///   from `prepare_queue`, an entry from request_queue should be moved over promptly.)
```

Reply:
@matklad beautiful! I had pondered on some of these ideas while reasoning about the semantics of the prepare queue & request queue as well (how we'd impose an extra RTT on the client even though we have space in our queues, relaxing the prepare queue invariant to avoid this), but I held off on proposing those because it was increasing the scope of this change a fair bit. Note to self - _be bolder next time_! 😎 

> And the existance of this alterantive solution actually warms me up quite a bit towards the simple approach we do in this PR for now!

Also, I agree with the _perfect_ implementation that you proposed! However, I have a small confusion about the "simple approach" that you talked about. You are referring to https://github.com/tigerbeetle/tigerbeetle/pull/2104#discussion_r1686483466, right?

Reply:
No, the "simple approach" is "reject request before adding it to request queue", as opposed to "add request to request queue, but don't convert it to prepare until the checkpoint is confirmed"

Reply:
Ah got it, thank you for clarifying! 👍🏼